DR-1781 Make cloud storage metadata cloud agnostic

nmalfroy opened the pull request:
This PR abstracts out the notion of storage, regions and resource types into interfaces and adds Azure equivalent implementations

There's some interesting deserialization code in here for figuring out, based on a region string or a resource type string, if it is an azure or google cloud option.

nmalfroy committed:
Add common method for looking up potentially non-existent enum value

nmalfroy committed:
Make regions and resources an interface so multiple clouds are supported

nmalfroy committed:
Make storage object an interface so multiple clouds are supported

nmalfroy committed:
Fix all usages of StorageResource

pshapiro4broad commented on src/main/java/bio/terra/app/model/AzureCloudResource.java:
```diff
@@ -0,0 +1,33 @@
+package bio.terra.app.model;
+
+/**
+ * Google cloud resources used by TDR.
```
typo
```suggestion
 * Azure cloud resources used by TDR.
```

nmalfroy replied:
🤦

pshapiro4broad commented on src/main/java/bio/terra/app/model/AzureRegion.java:
```diff
@@ -0,0 +1,121 @@
+        return value;
+    }
+
+    public String toString() {
+        return value;
+    }
```
Now that we've switched to use `getValue()` is this still needed? Keeping it around means that someone might depend on `toString()` without realizing it.

nmalfroy replied:
I agree.  I'm thinking I'll add `getValue()` to the interfaces and use those instead

pshapiro4broad commented on src/main/java/bio/terra/app/model/AzureRegion.java:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    /**
+     * Differs from ofValue in that this does not fail if no value is found
```
The method is `valueOf()` not `ofValue()`. It also differs in that it use a case insensitive comparison. I think it's nice to link to methods in javadoc as long as it's not excessive. The IDE will complete it for you if you type `#name`.
```suggestion
     * Differs from {@link #valueOf(String)} in that this does not fail if no value is found, and uses
     * a case insensitive compare
```

nmalfroy replied:
Gah, you're totally right. (and I totally agree about @link...I've been accused of using it too much :) )

nmalfroy committed:
Fix typo

Co-authored-by: Phil Shapiro <pshapiro@broadinstitute.org>

pshapiro4broad commented on src/main/java/bio/terra/app/model/CloudResource.java:
```diff
@@ -0,0 +1,60 @@
+                    CloudResource.class);
+            }
+            String value = node.textValue();
+            CloudResource resource = EnumUtils.valueOfLenient(GoogleCloudResource.class, value);
```
I would expect the class type to come last, as in `ObjectMapper.readValue()`. Also if you have this helper method, do we also need `AzureRegion.fromName()`?

nmalfroy replied:
I can switch the order.  I'm not sure I follow about the `AzureRegion.fromName`?

pshapiro4broad replied:
It looked to me like `AzureRegion.fromName(value)` was the same as `EnumUtils.valueOfLenient(AzureRegion.class, value)`.

pshapiro4broad commented on src/main/java/bio/terra/app/model/CloudRegion.java:
```diff
@@ -0,0 +1,62 @@
+ */
+@JsonDeserialize(using = CloudRegion.CloudRegionDeserializer.class)
+public interface CloudRegion {
+    String name();
```
Might be nice to have a javadoc here to indicate what `name()` represents / is used for.

nmalfroy replied:
yeah, that's a good point. It's admittedly weird since it's an Enum implementing it and that's a built in method

pshapiro4broad commented on src/main/java/bio/terra/app/model/GoogleRegion.java:
```diff
@@ -82,8 +88,10 @@ public static GoogleRegion fromValueWithDefault(String text) {
     public static boolean matchingRegionWithFallbacks(List<StorageResource> storage, GoogleRegion region) {
-        for (StorageResource resource : storage) {
-            switch (resource.getCloudResource()) {
+        List<StorageResource> gcpStorage =
```
I would use `var` here, and maybe use plural.
```suggestion
        var gcpStorageResources =
```

nmalfroy replied:
whoa! I've never actually used var...time to start!

pshapiro4broad commented on src/main/java/bio/terra/app/model/GoogleRegion.java:
```diff
@@ -82,8 +88,10 @@ public static GoogleRegion fromValueWithDefault(String text) {
-        for (StorageResource resource : storage) {
-            switch (resource.getCloudResource()) {
+        List<StorageResource> gcpStorage =
+            storage.stream().filter(s -> s.getCloudPlatform() == CloudPlatform.GCP).collect(Collectors.toList());
```
It might be clearer to put the cast here
```suggestion
            storage.stream().filter(s -> s.getCloudPlatform() == CloudPlatform.GCP).map(r -> (GoogleCloudResource) r).collect(Collectors.toList());
```

nmalfroy replied:
Ah I see.  Yeah, I'd need to change the variable definition too

pshapiro4broad commented on src/main/java/bio/terra/app/model/GoogleRegion.java:
```diff
@@ -82,8 +88,10 @@ public static GoogleRegion fromValueWithDefault(String text) {
     public static boolean matchingRegionWithFallbacks(List<StorageResource> storage, GoogleRegion region) {
-        for (StorageResource resource : storage) {
-            switch (resource.getCloudResource()) {
+        List<StorageResource> gcpStorage =
+            storage.stream().filter(s -> s.getCloudPlatform() == CloudPlatform.GCP).collect(Collectors.toList());
+        for (StorageResource resource : gcpStorage) {
+            switch ((GoogleCloudResource) resource.getCloudResource()) {
```
Also it seems odd to collect and then iterate over the collection.
```suggestion
        for (var resource : storage) {
            if (resource.getCloudPlatform() != CloudPlatform.GCP) {
                continue:
            }
            switch ((GoogleCloudResource) resource.getCloudResource()) {
```

nmalfroy replied:
Yeah, that's a really good point...I'll rework this bit

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/AzureStorageResource.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    @Override
+    public CloudResource getCloudResource() {
```
Can return a subtype here, due to java's covariant return type support.
```suggestion
    public AzureCloudResource getCloudResource() {
```

nmalfroy replied:
I don't so. I tried originally and it got really mad at me.  I suppose I could by adding generics to the interface...that would avoid having to use casts, which would be nice

pshapiro4broad replied:
What errors did you get? I checked out your branch and added this change and it seems to work, although I didn't run all the tests.

nmalfroy replied:
Ah, it breaks in the setter

nmalfroy replied:
Ah, fun with generics fixes this!  I'll push up some changes in a few minutes

pshapiro4broad replied:
You shouldn't need to use generics to fix this. Like I said, I checked out your code, made this change, and then compiled and started to run the tests. I could run all the tests if you think it will fail later.

pshapiro4broad replied:
> Ah, it breaks in the setter

Oh, so if the types don't match then the JSON serialization breaks? Oh well.

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/AzureStorageResource.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    @Override
+    public CloudRegion getRegion() {
```
Here too
```suggestion
    public AzureRegion getRegion() {
```

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/AzureStorageResource.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        AzureStorageResource that = (AzureStorageResource) o;
+        return Objects.equals(datasetId, that.datasetId) &&
+            cloudPlatform == that.cloudPlatform &&
+            cloudResource == that.cloudResource &&
+            region == that.region;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(datasetId, cloudPlatform, cloudResource, region);
+    }
+
+    @Override
+    public String toString() {
+        return "StorageResource{" +
+            "datasetId=" + datasetId +
+            ", cloudPlatform=" + cloudPlatform +
+            ", cloudResource=" + cloudResource +
+            ", region=" + region +
+            '}';
+    }
```
I think these can all be `default` in `StorageResource`

nmalfroy replied:
You can't...I tried :( It doesn't let you override `Object` methods...

pshapiro4broad replied:
That's a shame, but I guess it makes sense. I guess the next best thing is to add the code as `default` methods with slightly different names, and the subclass implementations can forward to the `default` ones.

pshapiro4broad replied:
e.g. 
```
    default boolean defaultEquals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        StorageResource that = (StorageResource) o;
        return Objects.equals(getDatasetId(), that.getDatasetId()) &&
            getCloudPlatform() == that.getCloudPlatform() &&
            getCloudResource() == that.getCloudResource() &&
            getRegion() == that.getRegion();
    }

    default int defaultHashCode() {
        return Objects.hash(getDatasetId(), getCloudPlatform(), getCloudResource(), getRegion());
    }

    default String defaultToString() {
        return "StorageResource{" +
            "datasetId=" + getDatasetId() +
            ", cloudPlatform=" + getCloudPlatform() +
            ", cloudResource=" + getCloudResource() +
            ", region=" + getRegion() +
            '}';
    }
```

nmalfroy replied:
Fair enough

nmalfroy replied:
ugh...I get warning on equals...gonna leave these alone for now

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/AzureStorageResource.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    @Override
+    public AzureStorageResource datasetId(UUID datasetId) {
+        this.datasetId = datasetId;
+        return this;
+    }
```
Do we need fluent style setters for this class? I'd really like to migrate away from mutable state in our objects in general, and switch to using builder-style construction or multi-arg constructors.

nmalfroy replied:
That's a broader question.  I agree but if we do that, I'd like to do a wholesale switchover

pshapiro4broad replied:
I disagree about the wholesale requirement, especially in a case like this where the change is specific to an individual class. In this case it's a new class in this PR, so it's an even better time to try doing things a different way IMO.

nmalfroy replied:
I think if we start using builders, I'd like to do it at a larger scale and also figure out what details about builder we want (e.g. separate class, same class, etc.)

pshapiro4broad replied:
In this case I don't think a builder is needed, as you always have all the arguments when you construct one. I think most cases where we use fluent-style setters could be written as constructor calls. There are probably a few cases where there's enough variation in construction that a builder would make sense, but in most cases I would bet we don't need to support all the combinations of setting one field or another.

pshapiro4broad replied:
I was thinking, another reason why the field should be `final` is because this class has hashCode(). If you make a collection with this class and then modify it, its hashcode and equality will change out from under the collection and its state could become invalid. In general we only want to make hashed (or ordered) collections over immutable data. (Unless the collection is aware of the mutation and can reorganize itself if its items state changes.)

pshapiro4broad reviewed (commented):
First round of comments

pshapiro4broad commented on src/main/java/bio/terra/app/model/CloudRegion.java:
```diff
@@ -0,0 +1,62 @@
+                    CloudRegion.class);
+            }
+            String value = node.textValue();
+            CloudRegion region = EnumUtils.valueOfLenient(GoogleRegion.class, value);
```
This system depends on no name collisions in the future. It would be nice to not have to add a namespace-type prefix to every enum element to prevent name collsions. This problem is often solved by having the type written as a separate field in the data, is that not possible here, e.g. if we're deserializing region and resource, do we always have a platform as well? If we did, we could use it to control this code.

nmalfroy replied:
We could.  FWIW, this is many many years away from being a problem (likely never).  Azure, GCP and AWS all have different naming schemes for their regions...so maybe one day we'll support Oracle cloud and they collide...but that seems...distant

nmalfroy committed:
fix javadoc

Co-authored-by: Phil Shapiro <pshapiro@broadinstitute.org>

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/StorageResource.java:
```diff
@@ -1,92 +1,47 @@
-    @Override
-    public int hashCode() {
-        return Objects.hash(datasetId, cloudPlatform, cloudResource, region);
+    static GoogleStorageResource getGoogleInstance() {
```
What is the benefit of having this method vs calling `new GoogleStorageResource()` directly?

nmalfroy replied:
Ah, great question.  I was doing this as a way to avoid the case where we would could create, say, a google resource and then set its cloud type to Azure.  I think given that we could just have it return a constant, I'll get rid of this static initializer

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/StorageResource.java:
```diff
@@ -1,92 +1,47 @@
+    default StorageResourceModel toModel() {
+        return new StorageResourceModel()
+            .cloudPlatform(getCloudPlatform())
+            .cloudResource(getCloudResource().toString())
+            .region(getRegion().toString());
```
Should these be `getValue()`?

nmalfroy replied:
Yeah...I'm going to add getValue to the interface

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/StorageResourceDao.java:
```diff
@@ -73,11 +75,18 @@ public StorageResourceDao(DataRepoJdbcConfiguration jdbcConfiguration) {
-                .cloudPlatform(CloudPlatform.valueOf(rs.getString("cloud_platform")))
-                .cloudResource(GoogleCloudResource.valueOf(rs.getString("cloud_resource")))
-                .region(GoogleRegion.valueOf(rs.getString("region")));
+            final CloudPlatform cloudPlatform = CloudPlatform.valueOf(rs.getString("cloud_platform"));
+            switch (cloudPlatform) {
```
It's a shame that `CloudPlatform` is auto generated, this would be a great case for a enum method.

pshapiro4broad commented on src/main/java/bio/terra/service/resourcemanagement/ResourceService.java:
```diff
@@ -77,7 +77,8 @@ public GoogleBucketResource getOrCreateBucketForFile(String possibleBucketName,
         String flightId) throws InterruptedException {
         final String datasetName = dataset.getName();
-        final GoogleRegion region = dataset.getDatasetSummary().getStorageResourceRegion(GoogleCloudResource.FIRESTORE);
+        final GoogleRegion region =
+            (GoogleRegion) dataset.getDatasetSummary().getStorageResourceRegion(GoogleCloudResource.FIRESTORE);
```
Is this cast necessary? It looks like you could get rid of it, if you change all the code under this to use `CloudRegion` instead of `GoogleRegion`.

nmalfroy replied:
I do need the cast since the underlying methods can return both

pshapiro4broad replied:
But in this case every method called downstream can be converted to accept `CloudRegion`. So the code can be changed here to use `CloudRegion` instead of depend on `GoogleRegion`. This is because none of the code requires a feature that's not part of the i`CloudRegion` interface.

pshapiro4broad commented on src/main/java/bio/terra/service/resourcemanagement/ResourceService.java:
```diff
@@ -105,7 +106,7 @@ public GoogleBucketResource getOrCreateBucketForFile(String possibleBucketName,
 
         return bucketService.getOrCreateBucket(bucketName, projectResource,
-            dataset.getDatasetSummary().getStorageResourceRegion(GoogleCloudResource.BUCKET),
+            (GoogleRegion) dataset.getDatasetSummary().getStorageResourceRegion(GoogleCloudResource.BUCKET),
```
Unlike the case above, this one ends up needing a `GoogleRegion`.

nmalfroy replied:
Right, because this is a google specific API

pshapiro4broad replied:
The more specific type is only needed if the contract is different. If you moved the google contracts to the base class (interface) then you could use the base class type everywhere. The code would throw at runtime if it had the wrong typed object, which is the equivalent of getting a type mismatch exception here.

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/AzureStorageResource.java:
```diff
@@ -0,0 +1,88 @@
+    private AzureRegion region;
+
+    AzureStorageResource() {
+        this.cloudPlatform = CloudPlatform.AZURE;
```
This could be done using a field initializer, and the field can be `final`. Or, you can remove the field and change `getCloudPlatform()` to `return CloudPlatform.AZURE;`

nmalfroy replied:
Yeah, that's a fair point...and we don't need it in the equality check since it's bound to the class type

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/StorageResourceDao.java:
```diff
@@ -96,7 +105,8 @@ public void createStorageAttributes(List<StorageResource> storageResources, UUID
                 regionParam, cloudResourceParam, platformParam));
 
-            params.addValue(regionParam, storageResource.getRegion().name());
+            params.addValue(regionParam, ((GoogleRegion) storageResource.getRegion())
```
Is this cast necessary? `parms.addValue()` accepts `Object`.

nmalfroy replied:
It's not. I actually ended up removing it in a second PR I'm working on...I'll remove it here

pshapiro4broad commented on src/test/java/bio/terra/app/model/CloudRegionTest.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Test
+    public void testDeserializeBadRegionValue() {
+        assertThat(
+            assertThrows(
+                InvalidFormatException.class,
+                () -> objectMapper.readValue("\"EUROPE_WEST100\"", CloudRegion.class),
+                "bad regions don't deserialize").getMessage(),
+            containsString("Unrecognized CloudRegion: EUROPE_WEST100"));
```
I don't think you need to wrap `assertThrows()` in `assertThat()`
```suggestion
        assertThrows(
            InvalidFormatException.class,
            () -> objectMapper.readValue("\"EUROPE_WEST100\"", CloudRegion.class),
            "bad regions don't deserialize").getMessage(),
        containsString("Unrecognized CloudRegion: EUROPE_WEST100");
```

pshapiro4broad replied:
OK, I think I see what this is doing.. but there must be a way to do it using `assertThrows()` directly..

nmalfroy replied:
Sadly, I don't think there is....I mean, aside from using assertJ (which does support it)

pshapiro4broad commented on src/test/java/bio/terra/common/fixtures/JsonLoader.java:
```diff
@@ -30,6 +31,11 @@ public String loadJson(final String resourcePath) throws IOException {
         return objectMapper.readerFor(resourceClass).readValue(json);
     }
 
+    public <T> T loadObject(final String resourcePath, final TypeReference<T> typeReference) throws IOException {
```
I may be crazy, but I think you can replace both of these `loadObject()` methods with a single method that infers its return type from its context. Something like this should work:
```
    public <T> T loadObject(String resourcePath) throws IOException {
        return objectMapper.readValue(loadJson(resourcePath), new TypeReference<T>() { });
    }
```

pshapiro4broad commented on src/test/java/bio/terra/service/dataset/DatasetDaoTest.java:
```diff
@@ -236,7 +236,7 @@ public void datasetTest() throws Exception {
 
             for (GoogleCloudResource resource: GoogleCloudResource.values()) {
-                GoogleRegion region = fromDB.getDatasetSummary().getStorageResourceRegion(resource);
+                GoogleRegion region = (GoogleRegion) fromDB.getDatasetSummary().getStorageResourceRegion(resource);
```
I think these cases can be converted to use `CloudRegion`, so the cast can be avoided.

pshapiro4broad reviewed (commented):
A few more comments

pshapiro4broad commented on src/main/java/bio/terra/app/model/CloudRegion.java:
```diff
@@ -0,0 +1,71 @@
+        }
+
+        @Override
+        public CloudRegion deserialize(
```
Have you heard of jackson support for deduction based polymorphism? I read about a while back, maybe it can help here? https://cowtowncoder.medium.com/jackson-2-12-most-wanted-1-5-deduction-based-polymorphism-c7fb51db7818

nmalfroy replied:
I got the current approach from the docs but I'll look here too...might be a cleaner way to do it

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/StorageResourceDao.java:
```diff
@@ -71,32 +73,40 @@ public StorageResourceDao(DataRepoJdbcConfiguration jdbcConfiguration) {
 
         List<String> valuesList = new ArrayList<>();
-        for (StorageResource storageResource : storageResources) {
+        for (StorageResource<?, ?> storageResource : storageResources) {
```
improve with `var`
```suggestion
        for (var storageResource : storageResources) {
```

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/StorageResource.java:
```diff
@@ -1,92 +1,40 @@
+    @JsonSubTypes.Type(value = AzureStorageResource.class, name = "azure"),
+    @JsonSubTypes.Type(value = GoogleStorageResource.class, name = "gcp")
+})
+public interface StorageResource<Region extends CloudRegion, Resource extends CloudResource> {
```
I'm not sure this is a great idea, to make this a generic type.

That being said, if you do this, you should make this abstract and move all the fields back here, except for `CloudPlatform` which would still be an `abstract` method with an override the subclasses.

You can use the type parameters to the class to declare the fields, the constructor args, and the getter/setter types.

pshapiro4broad replied:
Oh, and if you do that, then you can definitely move equals/tostring/hashcode back to `StorageResource`

nmalfroy replied:
Huh..interesting..let me play with that

nmalfroy replied:
And fwiw, without generics, it wasn't even test failing but straight up not compiling

nmalfroy replied:
Ugh...I tried the abstract route but...enums cant extend classes so no luck.  So we're either doing generics or putting casts everywhere...(unless there's another option I'm missing)

nmalfroy replied:
Wait scratch that last comment...wrong class

nmalfroy committed:
Address PR commnents

- Convert storage to generics
- Add getValue to the interface that the enums extend
- Add some javadoc
- StorageResource is now an abstract class
- Got rid of useless static creators
- Use a var in the GoogleRegion toModel method

tlangs commented on src/main/java/bio/terra/app/controller/SnapshotsApiController.java:
```diff
@@ -183,7 +183,9 @@ private AuthenticatedUserRequest getAuthenticatedInfo() {
     ) {
+        logger.info("Get auth");
         iamService.verifyAuthorization(getAuthenticatedInfo(), IamResourceType.DATASNAPSHOT, id, IamAction.READ_DATA);
+        logger.info("Get model");
```
Should these still be in here? If so, can we make them `debug`, and make the messages a bit more descriptive, like "Getting auth for snapshot with id :id"?

nmalfroy replied:
Ah, I'm getting rid of that commit...that was part of the integration test fixes.  I'll remove that commit

tlangs commented on src/main/java/bio/terra/app/model/AzureRegion.java:
```diff
@@ -0,0 +1,123 @@
+     * Differs from {@link #valueOf(String)} in that this does not fail if no value is found, and uses
+     * a case insensitive compare
+     * @param text region to look up
+     * @return A GoogleRegion or null if none is found
```
```suggestion
     * @return An AzureRegion or null if none is found
```

nmalfroy replied:
derp

nmalfroy committed:
Fix typo in javadoc

snf2ye reviewed (approved):
LGTM!
